### ISSUE TITLE
Memory leak on CPU

### ISSUE THREAD

**Issue by daniegr** [jacobgil/pytorch-grad-cam#78]
I am experiencing accumulating memory (i.e., memory leak) while performing GradCAM (v1.2.3) with a lightweight PyTorch model on Mac. Do you have any recommended solutions to combat this behavior when performing GradCAM on many different inputs sequentially?

**Comment by jacobgil** [comment]
Thanks for reporting this!
Can you please provide a bit more details on the scenario (and ideally a code snippet):
Do you have the same CAM object, and you're calling it on
cam(input_tensor=input_tensor, target_category=target_category)
in a loop?

I'm trying to reproduce the memory leak without success so far.

**Comment by daniegr** [comment]
@jacobgil: See the sketch of the overall loop below. I am iterating through the data points in eval_data, and want to perform GradCAM if the sample is of class 1. After computing GradCAM results are stored to file.
    
    # Construct model
    model = 
    model.eval()
    target_layer = model.features
    
    # Fetch evaluation data
    eval_data = load_data(data_path, labels_path)
    process = tqdm(eval_data)
  
    with open('cam.pkl', 'ab') as f:
        for batch_idx, (data, label, index) in enumerate(process):
            data = Variable(data.float(), requires_grad=False)
            label = Variable(label.long(),requires_grad=False)
  
            for data_index in range(data.shape[0]):
                if label[data_index,...] == 1:
                    cam = GradCAM(model=model, target_layer=target_layer, use_cuda=False)
                    cam = cam(input_tensor=torch.unsqueeze(data[data_index,...], 0), target_category=1)
                    cam = np.mean(cam, axis=0)
                    pickle.dump(cam, f)

**Comment by jacobgil** [comment]
Thanks.
The cam object,
`cam = GradCAM(model=model, target_layer=target_layer, use_cuda=False)`
Should be created only once - it's re-used between all images.
I verified that with it the memory grows, so it's probably this.

**Comment by jacobgil** [comment]
Btw, I noticed
`cam = np.mean(cam, axis=0)`
And just wanted to make sure it's intentional. The cam image returned probably is a spatial image of size 224x224, so the mean on axis =0 is taking the mean on the rows.

**Comment by daniegr** [comment]
Great, @jacobgil, this combats the memory leak! I apologize for the inappropriate use of your framework, and I suggest you to specify in the documentation that the GradCAM object should only be created once :) Regarding the apparently strange operation at the end there it was intentional since I am employing your framework on graph convolutional networks for skeleton-based action recognition, where we want discriminative values of the _N_ body parts across _T_ time steps, where the spatial map is _T_ x _N_, hence the mean operation. Thank you for this great codebase!
